### PR TITLE
Issue/2/new instance return value

### PR DESCRIPTION
**Pull request by ilyavf**
@justinbmeyer  Pass a callback and arguments so that an alternative value could be created for a new instance.

### CODE REVIEW

**Comment by ilyavf** [comment]
Fixed acc to feedback. Need to add docs...

File: can-construct.js [hunk 1]
```diff
@@ -526,6 +529,49 @@ assign(Construct, {
...
+	 * ```
+	 * const myStore = {};
+	 * 
+	 * const Item = Construct.extend({
+	 *     setup: function(params){
+	 *         if (myStore[params.id]){
+	 *             var item = myStore[params.id];
+	 *             
+	 *             // Merge new data to the existing instance:
+	 *             Object.keys( params ).forEach(function( key ){
```
Comment (justinbmeyer) [inline can-construct.js]: Could probably use `Object.assign` here.

File: can-construct.js [hunk 2]
```diff
@@ -526,6 +529,49 @@ assign(Construct, {
...
+	 *             });
+	 *             
+	 *             // Return the updated item:
+	 *             return new Construct.ReturnValue( item );
+	 *         } else {
+	 *             return [params];
+	 *         }
+	 *     },
+	 *     init: function(params){
+	 *         this.id = params.id;
```
Comment (justinbmeyer) [inline can-construct.js]: Might use `Object.assign` here too.

File: can-construct.js [hunk 3]
```diff
@@ -526,6 +529,49 @@ assign(Construct, {
...
+	 *         this.name = params.name;
+	 *         
+	 *         // Save to cache store:
+	 *         myStore[this.id] = this;
```
Comment (justinbmeyer) [inline can-construct.js]: This should almost certainly be done in `setup`.  `init` is usually left for people inheriting from `Item` to use for their particular `Item` type.

File: can-construct.js [hunk 4]
```diff
@@ -663,6 +709,40 @@ assign(Construct, {
 		 * console.log(Counter.count); // 1
 		 * ```
 		 */
+	},
+	/**
+	 * @function can-construct.ReturnValue ReturnValue
+	 * @parent can-construct.static
+	 * 
+	 * A constructor function which `value` property will be used as a value for a new instance.
```
Comment (justinbmeyer) [inline can-construct.js]: Too much detail for a description. It shouldn't care at all about the `value` property. Maybe something like:

> Use to overwrite the return value of `new Constructor(...)`.

File: can-construct.js [hunk 5]
```diff
@@ -663,6 +709,40 @@ assign(Construct, {
...
 		 */
+	},
+	/**
+	 * @function can-construct.ReturnValue ReturnValue
+	 * @parent can-construct.static
+	 * 
+	 * A constructor function which `value` property will be used as a value for a new instance. 
+	 * 
+	 * @signature `new Construct.ReturnValue( value )`
+	 *
```
Comment (justinbmeyer) [inline can-construct.js]: The signature should talk about how this works.  Also, the example code below should be put in the signature.

File: can-construct.js [hunk 6]
```diff
@@ -663,6 +709,40 @@ assign(Construct, {
...
+	},
+	/**
+	 * @function can-construct.ReturnValue ReturnValue
+	 * @parent can-construct.static
+	 * 
+	 * A constructor function which `value` property will be used as a value for a new instance. 
+	 * 
+	 * @signature `new Construct.ReturnValue( value )`
+	 * 
+	 * 	@param {*} value A value to be used for a new instance instead of a new object.
```
Comment (justinbmeyer) [inline can-construct.js]: I think the `value` MUST be an Object type.  If someone tries to return a number, it won't work.

File: can-construct.js [hunk 7]
```diff
@@ -675,8 +755,9 @@ assign(Construct, {
  *
  * @param {*} args The arguments passed to the constructor.
  *
- * @return {Array|undefined} If an array is returned, the array's items are passed as
- * arguments to [can-construct::init init]. The following example always makes
+ * @return {Array|undefined|can-construct.ReturnValue} If an array is returned, the array's items are passed as
+ * arguments to [can-construct::init init]. If [can-construct.ReturnValue] is returned then the value that passed to it
+ * will be used as a value for a new instance. The following example always makes
```
Comment (justinbmeyer) [inline can-construct.js]: Change:

> If [can-construct.ReturnValue] is returned then the value that passed to it
> will be used as a value for a new instance.

To something more like:

If a [can-construct.ReturnValue] instance is returned, the `ReturnValue` instance's 
value will be returned as the result of calling `new Constructor()`.

File: can-construct.js (hunk 6 above)
Comment (justinbmeyer) [inline can-construct.js]: Also, indent 2 spaces `@param` and other nested `@` tags.

File: can-construct.js (hunk 1 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

File: can-construct.js (hunk 2 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

File: can-construct.js (hunk 3 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

File: can-construct.js (hunk 5 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

File: can-construct.js (hunk 6 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

File: can-construct.js (hunk 7 above)
Comment (ilyavf) [inline can-construct.js]: Fixed.

**Pull request merged**